[GT-183] Add service types page for non-admins

Sae126V opened the pull request:
- Admin => Can do CRUD operations in `Service Types` page.
- non-Admin => Can ONLY read the data(view) in `Service Types` page.

"Resolves #126 " 

Note:
Sometimes GIT interface will not show the EOL. If the file is NOT having EOL GIT may shows `no_newline_at_end` explicitly.

Sae126V commented:
Ready for the review of Functionality. @gregcorbett

gregcorbett commented:
Looks mostly good - a couple of things I spotted were:

- `Page_Type=Service_Types` doesn't display for non-registered users, so there will be an bug somewhere.
- Move "ServiceTypes" to the browse section for admins, and position it between "Scopes" and "Role Action Map" for both admins and non admins.
- Under GOCDB Admins, have an "Add Service Type" link leading to `Page_Type=Admin_Add_Service_Type`, and position it after "Add Project"

Sae126V commented:
> Looks mostly good - a couple of things I spotted were:
> 
> * `Page_Type=Service_Types` doesn't display for non-registered users, so there will be an bug somewhere.
> * Move "ServiceTypes" to the browse section for admins, and position it between "Scopes" and "Role Action Map" for both admins and non admins.
> * Under GOCDB Admins, have an "Add Service Type" link leading to `Page_Type=Admin_Add_Service_Type`, and position it after "Add Project"

I have added a single page intentionally to avoid showing the same content in both, Greg. But with your suggestion the problem is not valid anymore. Updated the code to achieve the desired functionality.

Sae126V commented:
Ready to be viewed for functionality.

gregcorbett commented:
> * `Page_Type=Service_Types` doesn't display for non-registered users, so there will be an bug somewhere.

ah the same thing happens with `Page_Type=Service_Type`, there's probably a similar bug.

> Move "ServiceTypes" to the browse section for admins, and position it between "Scopes" and "Role Action Map" for both admins and non admins.

Ah sorry, I meant between "Services" and "Scopes".

Sae126V commented:
> > * `Page_Type=Service_Types` doesn't display for non-registered users, so there will be an bug somewhere.
> 
> ah the same thing happens with `Page_Type=Service_Type`, there's probably a similar bug.
> 
> > Move "ServiceTypes" to the browse section for admins, and position it between "Scopes" and "Role Action Map" for both admins and non admins.
> 
> Ah sorry, I meant between "Services" and "Scopes".

Cool. Updated the flow, Greg.

gregcorbett commented:
A previous version I looked at had the "Add Service Type" functionality on `Page_Type=Service_Types` for admins, please add this back in.

Sae126V commented:
> A previous version I looked at had the "Add Service Type" functionality on `Page_Type=Service_Types` for admins, please add this back in.

The functionality got changed from previous version, Greg. 

I have added the below one.
Under GOCDB Admins, have an "Add Service Type" link leading to Page_Type=Admin_Add_Service_Type, and position it after "Add Project"

Since, we are providing an admin option to directly goto the Admin_Add_Service_Type page Just like Add Projects, Add NGI. I have implemented the same for Add service Type.
 
And "Add Service Type" functionality on `Page_Type=Service_Types` is no longer needed for normal users or admin. Since admin has the page separately for adding service types. As it only admin rights. They can add service type from admin section. Does that make sense?

gregcorbett commented:
> > A previous version I looked at had the "Add Service Type" functionality on `Page_Type=Service_Types` for admins, please add this back in.
> 
> The functionality got changed from previous version, Greg.
> 
> I have added the below one. Under GOCDB Admins, have an "Add Service Type" link leading to Page_Type=Admin_Add_Service_Type, and position it after "Add Project"
> 
> Since, we are providing an admin option to directly goto the Admin_Add_Service_Type page Just like Add Projects, Add NGI. I have implemented the same for Add service Type.

Fair enough, that does give conistency with the Site, NGI etc pages.

Functionality looks good to me 😄

gregcorbett commented on htdocs/web_portal/controllers/admin/view_service_type.php:
```diff
@@ -24,31 +24,32 @@
+    $identifier = Get_User_Principle();
+    $user = \Factory::getUserService()->getUserByPrinciple($identifier);
+
+    /**
```
We can take this oppurtunity to remove this comment, it doesn't add any additional context to the code IMO.

gregcorbett commented on htdocs/web_portal/views/admin/added_service_type.php:
```diff
@@ -4,7 +4,7 @@
     <?php require_once __DIR__ . '/../fragments/serviceTypeInfo.php'; ?>
 
-    <a href="index.php?Page_Type=Admin_Service_Type&amp;id=<?php echo $params['ID'] ?>">
+    <a href="index.php?Page_Type=Service_Type&amp;id=<?= $params['ID']; ?>">
```
why isn't the php tag and echo needed here? I can see that it works, but I'm not at all sure why it works.

Sae126V replied:
Ahh. It was exceeding 80. So I have used another short version for  <?php echo i.e <?= ?> :)

gregcorbett reviewed (changes requested):
I've left some comments.

Can you move `{controllers, views/view_service_type(s).php}` out of the relevant admin folders and into new `service_type` folders?

Sae126V commented on config/web_portal/menu.xml:
```diff
@@ -42,6 +42,12 @@
+        <show_on_instance>write_enabled</show_on_instance>
+        <name>Service Types</name>
+        <link>index.php?Page_Type=Service_Types</link>
+    </ViewServiceTypes>
```
Make sense. Updated the code

Sae126V commented on htdocs/web_portal/controllers/utils.php:
```diff
@@ -817,3 +817,13 @@ function getReadPDParams($user)
+
+/**
+ * Helper to identify whether the user is an Admin or NOT.
+ *
```
to differentiate a bit from checkUserIsAdmin()
-        I have updated the comments. 

I do think both methods are needed.
-      I am not sure what is referring to Greg. So NOT commenting on that.

gregcorbett replied:
at first I wondered why both `checkUserForAdminCredentials` and `checkUserIsAdmin` are needed - given one might think they do similar things.

It looks to me like `checkUserIsAdmin` is more suitable for restricting access to whole pages, and `checkUserForAdminCredentials` is more suitable for deciding whether to display elements.

Sae126V replied:
Cool. Yep make sense, Greg.

gregcorbett commented on htdocs/web_portal/views/admin/view_service_types.php:
```diff
@@ -37,16 +24,23 @@
-                                </a>
+                                <?php
+                                echo "<a href=\"index.php?Page_Type="
+                                    . $searchName
```
The Page_Type, i.e. where we want this link to go, is always `Service_Type` though.

Sae126V replied:
I found one interesting problem here, Greg :). So, Whenever possible, I would like to show this scenario to you. Shall we discuss this at some time ?

Sae126V replied:
As discussed, Updated the code in 0de0b6a887fc6c4c76d65a44e27064ff4dde5961.

gregcorbett commented:
Looks good to me, cna you squash the 4 commits into the first one please?

Sae126V commented:
I squashed into one @gregcorbett . Their is styles PR for this one so can we park it for a while so that I can raise a styles PR ?

gregcorbett commented:
rebasing on latest dev
